colexec: make hash table more dynamic and fix hash join in some cases

yuzefovich opened the pull request:
**colexec: make hash table more dynamic**

This commit replaces all fixed allocations of constant size in the hash
table in favor of allocating the slices dynamically which makes the hash
table more dynamic as well. This allows us to place the logic of
clearing up those slices for the next iteration in a single place which
simplifies the code a bit.

Additionally it moves `buckets` slice that is only used by the hash
joiner out of the hash table.

Release note: None

**colexec: fix LEFT ANTI hash join when right eq cols are key**

Release note (bug fix): Previously, CockroachDB could return incorrect
results when performing LEFT ANTI hash join when right equality columns
form a key when it was executed via the vectorized engine, and now this
has been fixed.

**colexec: fix set-op hash joins and optimize them a bit**

This commit fixes several problems with set-operation hash joins:
1. similar to how LEFT ANTI hash joins are handled, INTERSECT ALL and
EXCEPT ALL hash joins rely on `headID` to be populated. That step is
skipped if the right side is distinct (meaning right equality columns
form a key). This would result in incorrect output, and we now override
`rightDistinct` flag to get the desired behavior (probably sub-optimal
but correct).
2. actually, before we would get an incorrect result, we would hit an
out of bounds error because `hashTable.visited` would not have been
created previously. That slice is used by set-op joins to track which
tuples from the right side have already been "deleted" (i.e. they were
used for a match). This is now also fixed.

However, currently the information whether the right equality columns
form a key is not propagated for set-operation joins, so the bug would
not have occurred in the non-test environment.

Additionally, this commit optimizes the handling of LEFT ANTI and EXCEPT
ALL joins by not allocating `same` slice because those two join types
have separate `collectAnti*` methods that don't use that slice.

Release note: None

yuzefovich committed:
colexec: make hash table more dynamic

This commit replaces all fixed allocations of constant size in the hash
table in favor of allocating the slices dynamically which makes the hash
table more dynamic as well. This allows us to place the logic of
clearing up those slices for the next iteration in a single place which
simplifies the code a bit.

Additionally it moves `buckets` slice that is only used by the hash
joiner out of the hash table.

Release note: None

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but I think it'd still be good to include a description of the cleanup in the commit message

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53226#-:-MFH_Bes8KzQ39YXAEdR:bli3r3l)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)



<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexec: fix set-op hash joins and optimize them a bit

This commit fixes several problems with set-operation hash joins:
1. similar to how LEFT ANTI hash joins are handled, INTERSECT ALL and
EXCEPT ALL hash joins rely on `headID` to be populated. That step is
skipped if the right side is distinct (meaning right equality columns
form a key). This would result in incorrect output, and we now override
`rightDistinct` flag to get the desired behavior (probably sub-optimal
but correct).
2. actually, before we would get an incorrect result, we would hit an
out of bounds error because `hashTable.visited` would not have been
created previously. That slice is used by set-op joins to track which
tuples from the right side have already been "deleted" (i.e. they were
used for a match). This is now also fixed.

However, currently the information whether the right equality columns
form a key is not propagated for set-operation joins, so the bug would
not have occurred in the non-test environment.

Additionally, this commit optimizes the handling of LEFT ANTI and EXCEPT
ALL joins by not allocating `same` slice because those two join types
have separate `collectAnti*` methods that don't use that slice.

Release note: None

yuzefovich committed:
colexec: fix LEFT ANTI hash join when right eq cols are key

Release note (bug fix): Previously, CockroachDB could return incorrect
results when performing LEFT ANTI hash join when right equality columns
form a key when it was executed via the vectorized engine, and now this
has been fixed.

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but does it make sense to squash the second and third commit? The third commit implies that the second one could still produce a panic by itself.

> Reviewed 8 of 8 files at r3, 2 of 2 files at r4, 6 of 6 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53226#-:-MFU70qXDpt0pKUBNeWm:byqy9ky)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colexec/hashjoiner.go, line 674 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/53226#-MFU5wp1DslZ0FueS_p3:-MFU5wp1DslZ0FueS_p4:bhoot6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/c50ccc998440a888620ff005388bfb623a911ef6/pkg/sql/colexec/hashjoiner.go#L674)):*
> ```Go
> 		// matching. However, headID is only populated when the right side is
> 		// considered to be non-distinct, so we override that information here.
> 		rightDistinct = false
> ```

This feels forced because to me it seems that we're setting a seemingly unrelated variable to enable some required behavior (indeed, it seems to be the opposite of what you'd expect with right equality columns forming a key). Consider introducing a more explicit way to enable the behavior you want. Not a requirement for this PR to merge.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I specifically carved out the second commit into a separate one so that it was easier to backport it (set-op joins were introduced in 20.2 release, but Left Anti is present in 20.1). And the implication of the third commit is slightly different - the panic would occur only with set-op joins, but the optimization of not allocating `same` slice is relevant for Left Anti and Except All (I'm not backporting the optimization).

TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53226#-:-MFW3Tbg9ptLm1xMDB-b:bqxeyt7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colexec/hashjoiner.go, line 674 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/53226#-MFU5wp1DslZ0FueS_p3:-MFW38lmCAJMB6NOjFQD:b-d3g3p8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c50ccc998440a888620ff005388bfb623a911ef6/pkg/sql/colexec/hashjoiner.go#L674)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This feels forced because to me it seems that we're setting a seemingly unrelated variable to enable some required behavior (indeed, it seems to be the opposite of what you'd expect with right equality columns forming a key). Consider introducing a more explicit way to enable the behavior you want. Not a requirement for this PR to merge.
</blockquote></details>

I agree, it's a bit "forced", but we already do a similar thing in Left Semi case above (just in the opposite direction), so I think it's ok. The reasoning is that we want the "non-distinctness" behavior to be triggered, regardless of the fact whether the right side is distinct or not.


<!-- Sent from Reviewable.io -->
